Keep id on paste if internal link points to it

juanruitina opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
With these changes, the block editor will keep the IDs in links when there's another link pointing to that ID. This is useful when pasting documents with internal links, e.g. when pasting text with footnotes from Microsoft Word, which used to work with the TinyMCE editor. Fixes #12784.

Sometimes internal links leverage the "name" attribute as the anchor instead of "id". However, the use of the `name` attribute for `a` elements [is obsolete in HTML5](https://html.spec.whatwg.org/multipage/obsolete.html#obsolete-but-conforming-features); use of `id` is recommended instead; [W3C Markup Validator](https://validator.w3.org/) also shows a warning. This code also saves the `name` attribute in links as the link's `id` (unless there's an `id` attribute already), and then removes it.

(Following @ellatrix feedback in #27677, these changes affect all links, not only footnotes and endnotes)

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
Tested locally with wp-env.
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->(Doesn't apply)

juanruitina committed:
Keep id on paste if internal link points to it

If link has obsolete name attribute, save as id and remove name
Remove links with no id nor href attributes (+ test)
Update raw handling readme: footnote support, LibreOffice

juanruitina commented:
@ellatrix, I accidentally removed the branch I used for #27677, so the PR closed. I took the chance to name the new branch following the developer guidelines. Apologies for the inconvenience.

ellatrix commented on packages/dom/src/phrasing-content.js:
```diff
@@ -32,7 +32,7 @@ const textContentSchema = {
 	del: {},
 	ins: {},
-	a: { attributes: [ 'href', 'target', 'rel' ] },
+	a: { attributes: [ 'href', 'target', 'rel', 'id', 'name' ] },
```
I think it's not necessary to add `name` here, because the removal of HTML only happens after the phrasing content reducer.

juanruitina replied:
I didn't know this, that's great. Just removed it.

ellatrix reviewed (approved):
I think we can remove `name` from the schema.

Otherwise this is great. Simple and generalised so that's really nice.

ellatrix commented:
Looks like there's something going wrong with the tests so you might have to rebase to re-run them again at a later time.

ellatrix commented on packages/blocks/src/api/raw-handling/phrasing-content-reducer.js:
```diff
@@ -56,5 +56,25 @@ export default function phrasingContentReducer( node, doc ) {
+		}
+
+		// Remove anchor tag if no href nor id
+		if ( ! node.id && ! node.href ) {
```
Hm, shouldn't this only unwrap only if there's no link _elsewhere_? Why would a node have both an id and href?

Also, I'm not sure if we should concern ourselves with removing nodes here. That's something the `removeInvalidHTML` filter will (or should) take care of.

juanruitina replied:
Understood. I removed this check as well, let's have `removeInvalidHTML` take care of this.

ellatrix commented on packages/blocks/src/api/raw-handling/phrasing-content-reducer.js:
```diff
@@ -56,5 +56,25 @@ export default function phrasingContentReducer( node, doc ) {
+			node.id = node.name;
+		}
+		// Removes obsolete name attribute in any case
+		node.removeAttribute( 'name' );
```
If we remove `name` from the schema, `removeInvalidHTML` should take care of this.

juanruitina committed:
Remove unnecessary code

No need to remove name attribute, handled by removeInvalidHTML

juanruitina commented:
Thanks a lot @ellatrix for your indications and suggestions, I incorporated them and rebased as suggested. It seems I need approval for rerunning tests, can you do that?

ellatrix commented:
@juanruitina You need to rebase to re-run them :)

ellatrix commented:
Oh, one last thing, could you add an paste integration test? I think they're in `test/integration`. You can put the MS Word data in an HTML file like the other tests.

juanruitina committed:
Add Word paste test for internal links

juanruitina commented on test/integration/fixtures/ms-word-out.html:
```diff
@@ -37,3 +37,43 @@ <h2>This is a heading level 2</h2>
+<!-- /wp:paragraph -->
+
+<!-- wp:paragraph -->
+<p><a>This is a paragraph with an anchor with no link pointing to it.</a></p>
```
@ellatrix Could you provide some more guidance on how could we unwrap text within links with no attributes (therefore with no use)? You mentioned removeInvalidHTML should handle it but I couldn't figure out how we could make sure it achieves that (it doesn't at the moment). It would be nice to clean it up a bit.

juanruitina commented:
I added a test in `test/integration` for the Word paste, seems to work.

(I did something wrong with the rebase and Github automatically asked for reviews from many code owners. I could undo the error, but not those review requests. Sorry for that!!! Geez, I need to get more familiar with git)

Mamaduka commented:
Thanks for working on this, @juanruitina, and congratulations on the first merged PR 🎉
